sdk/src/types/block/unlock/mod.rs

```diff
@@ -120,40 +133,60 @@ impl Unlocks {
     }
 }
 
+fn verify_unlock<'a>(
+    unlocks: &'a [Unlock],
+    unlock: &'a Unlock,
+    index: u16,
+    seen_signatures: &mut HashSet<&'a SignatureUnlock>,
+) -> Result<(), Error> {
+    match unlock {
+        Unlock::Signature(signature) => {
+            if !seen_signatures.insert(signature.as_ref()) {
+                return Err(Error::DuplicateSignatureUnlock(index));
+            }
+        }
+        Unlock::Reference(reference) => {
+            if index == 0
+                || reference.index() >= index
+                || !matches!(unlocks[reference.index() as usize], Unlock::Signature(_))
+            {
+                return Err(Error::InvalidUnlockReference(index));
+            }
+        }
+        Unlock::Account(account) => {
+            if index == 0 || account.index() >= index {
+                return Err(Error::InvalidUnlockAccount(index));
+            }
+        }
+        Unlock::Anchor(anchor) => {
+            if index == 0 || anchor.index() >= index {
+                return Err(Error::InvalidUnlockAnchor(index));
+            }
+        }
+        Unlock::Nft(nft) => {
+            if index == 0 || nft.index() >= index {
+                return Err(Error::InvalidUnlockNft(index));
+            }
+        }
+        Unlock::Multi(_) => return Err(Error::MultiUnlockRecursion),
```

Comment:
This makes sense if it's supposed to be only called inside of `verify_unlocks()`, but just looking at the function itself it's not clear why this would return a recursion error, so maybe add some comment here or above the function

Reply:
Added a comment and renamed the function